vllm/executor/ray_gpu_executor.py

```diff
@@ -234,6 +243,19 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
                           max_concurrent_workers=self.parallel_config.
                           max_parallel_loading_workers)
 
+        if self.use_ray_spmd_worker:
+            for pp_rank in range(self.parallel_config.pipeline_parallel_size):
+                self.pp_tp_workers.append([])
+                for tp_rank in range(
+                        self.parallel_config.tensor_parallel_size):
+                    # PP=2, TP=4
+                    # pp_tp_workers = [[0, 1, 2, 3], [4, 5, 6, 7]]
+                    rank = (pp_rank * self.parallel_config.tensor_parallel_size
+                            ) + tp_rank
+                    assert len(self.pp_tp_workers[pp_rank]) == tp_rank
+                    assert pp_rank < len(self.pp_tp_workers)
+                    self.pp_tp_workers[pp_rank].append(self.workers[rank])
```

Comment:
This could potentially lead to hang, `self.workers` may not be in rank order. See lines 268-275 in this file

Reply:
`self.workers` is sorted with `sort_by_driver_then_worker_ip` on L171. That should be sufficient?

Reply:
I might be understanding wrong but  AFAIU `sort_by_driver_then_worker_ip` ensures that the workers will be sorted by IP address. It does *not* ensure that the PyTorch rank of `workers` is in strictly increasing order.

Reply:
So after sorting, the order in `self.workers` will be used as ranks, see https://github.com/vllm-project/vllm/pull/6837/files#diff-4fc454dabd93d2c8a43a962360afe9753d139415326bc3097c99815bf0311736R251

Reply:
Yes, but the actual PyTorch rank may be different

Reply:
See: https://github.com/vllm-project/vllm/pull/6399

Reply:
Thanks for the pointers. Took a look at #6399 and #6235, I think the current implementation should be correct.

Without `sort_by_driver_then_worker_ip`, then we need to do something like #6399 and rely on `worker_rank`. But after `sort_by_driver_then_worker_ip`, for an example of PP=4, TP=2, let's say `self.workers` is:
[W0, W1, W2, W3, W4, W5, W6, W7],
then their ranks are [0, 1, 2, 3, 4, 5, 6, 7], these are what got used in `init_worker` and `init_device` (i.e., PyTorch rank). In other words, after sorting, the position can be used as the rank when initializing ADAG workers.

Please let me know if this helps with clarification.

Reply:
I think I understand now. Instead of initializing then getting pytorch ranks, we sort, then use the sorted list as the ranks. If this is the case, would `worker_ranks` then still even be relevant? I.e. this section: https://github.com/vllm-project/vllm/blob/4fd1525dce3284a44298d53f311c0bbdbf72a32c/vllm/executor/ray_gpu_executor.py#L198-L209.

It seems it's only used for `tp_driver_workers` and `non_driver_workers` now, if so, does it make sense to create these lists at the same time as `self.pp_tp_workers` and get rid of `worker_ranks` this would get rid of unnecessary duplication. Note this is different from the below comment, since we're still creating separate lists.

Reply:
I got rid of `worker_ranks` and it simplified the logic quite a bit! And all tests passed (Test Trace failure is irrelevant and probably will be fixed after rebase).

Reply:
Awesome!! Just a small thing you might want to run the _multinode_ tests a few times. This passed like 50% of the time, as the failures we saw before adding extra logic for `worker_ranks` were intermittent

Reply:
Thanks. After this update multinode tests have run at least three times and all passed. (Before this more rounds have passed). I will run the tests more.